### PR TITLE
feat: add code hash to storage account

### LINKED ISSUE #1267: feat: add code hash to the Account storage
### Describe the Feature Request

Currently the EXTCODEHASH opcode computes the hash of the deployed account.

We should cache this value so not to do it again and again.

### PR DESCRIPTION
**Pull request by obatirou**
Time spent on this PR:

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves #1267

## What is the new behavior?

- code hash is added to model.Account
- it is computed only once when creating a contract account

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot/1309)

### CODE REVIEW

File: src/kakarot/accounts/library.cairo [hunk 1]
```diff
@@ -436,6 +440,20 @@ namespace AccountContract {
         );
         return is_valid_jumpdest(index=index);
     }
+
+    func get_code_hash{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (
+        code_hash: Uint256
+    ) {
+        let (code_hash) = Account_code_hash.read();
+        return (code_hash,);
+    }
```
Comment (enitrat) [inline src/kakarot/accounts/library.cairo:449]:
```suggestion
    func get_code_hash{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> Uint256 {
        let (code_hash) = Account_code_hash.read();
        return code_hash;
    }
```

prefer returning non-tuple single values when functions are not entrypoints

File: src/kakarot/accounts/library.cairo [hunk 2]
```diff
@@ -436,6 +440,20 @@ namespace AccountContract {
...
+
+    func set_code_hash{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        code_hash: Uint256
+    ) -> () {
+        Account_code_hash.write(code_hash);
+        return ();
+    }
```
Comment (enitrat) [inline src/kakarot/accounts/library.cairo:456]:
```suggestion
    func set_code_hash{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
        code_hash: Uint256
    ) {
        Account_code_hash.write(code_hash);
        return ();
    }
```

empty return type can be omitted

File: src/kakarot/instructions/environmental_information.cairo [hunk 3]
```diff
@@ -11,15 +11,16 @@ from starkware.cairo.common.math_cmp import is_not_zero, is_nn
 from starkware.cairo.common.uint256 import Uint256, uint256_le
 
 from kakarot.account import Account
-from kakarot.interfaces.interfaces import ICairo1Helpers
+from kakarot.interfaces.interfaces import ICairo1Helpers, IAccount
```
Comment (enitrat) [inline src/kakarot/instructions/environmental_information.cairo:14]:
```suggestion
from kakarot.interfaces.interfaces import IAccount
```

unused

File: src/kakarot/instructions/environmental_information.cairo [hunk 4]
```diff
@@ -485,14 +486,8 @@ namespace EnvironmentalInformation {
...
-        let (implementation) = Kakarot_cairo1_helpers_class_hash.read();
-        let (code_hash) = ICairo1Helpers.library_call_keccak(
-            class_hash=implementation,
-            words_len=dst_len,
-            words=dst,
-            last_input_word=last_word,
-            last_input_num_bytes=last_word_num_bytes,
-        );
+        let (starknet_address) = Kakarot_evm_to_starknet_address.read(evm_address);
+        let (code_hash) = IAccount.get_code_hash(contract_address=starknet_address);
```
Comment (enitrat) [inline src/kakarot/instructions/environmental_information.cairo:490]:
```suggestion
        let code_hash = account.code_hash
 ```

I don't think a starknet call should be done here, the account and its bytecode should have been cached already at this point

File: src/kakarot/account.cairo [hunk 5]
```diff
@@ -340,6 +368,51 @@ namespace Account {
...
+            created=self.created,
+        );
+    }
+
+    // @notice Compute the code hash of an account
+    // @param address The address of the account
+    // @return the hash of the code
+    func compute_code_hash{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        self: model.Account*
+    ) -> (code_hash: Uint256) {
```
Comment (enitrat) [inline src/kakarot/account.cairo:398]:
```suggestion
    ) -> Uint256 {
```

File: src/kakarot/account.cairo [hunk 6]
```diff
@@ -340,6 +368,51 @@ namespace Account {
...
+
+        let (implementation) = Kakarot_cairo1_helpers_class_hash.read();
+        let (code_hash) = ICairo1Helpers.library_call_keccak(
+            class_hash=implementation,
+            words_len=dst_len,
+            words=dst,
+            last_input_word=last_word,
+            last_input_num_bytes=last_word_num_bytes,
+        );
+        return (code_hash=code_hash);
```
Comment (enitrat) [inline src/kakarot/account.cairo:413]:
```suggestion
        return code_hash;
```

File: src/kakarot/account.cairo [hunk 7]
```diff
@@ -126,14 +141,21 @@ namespace Account {
 
         let (bytecode_len, bytecode) = IAccount.bytecode(contract_address=starknet_address);
         let (nonce) = IAccount.get_nonce(contract_address=starknet_address);
+        IAccount.get_code_hash(contract_address=starknet_address);
```
Comment (enitrat) [inline src/kakarot/account.cairo:144]: is there specific reason for not doing `let (code_hash) = IAccount.get_code_hash(contract_address=starknet_address);` ?

File: src/backend/starknet.cairo [hunk 8]
```diff
@@ -35,6 +35,7 @@ from kakarot.storages import (
     Kakarot_base_fee,
     Kakarot_block_gas_limit,
     Kakarot_prev_randao,
+    Kakarot_cairo1_helpers_class_hash,
```
Comment (enitrat) [inline src/backend/starknet.cairo:38]:
```suggestion
```

File: src/kakarot/account.cairo [hunk 9]
```diff
@@ -21,18 +21,20 @@ from starkware.cairo.common.hash_state import (
 )
 from starkware.starknet.common.storage import normalize_address
 from starkware.starknet.common.syscalls import get_contract_address
-
+from starkware.cairo.lang.compiler.lib.registers import get_fp_and_pc, get_ap
```
Comment (enitrat) [inline src/kakarot/account.cairo:24]:
```suggestion
from starkware.cairo.lang.compiler.lib.registers import get_ap
```

File: src/kakarot/accounts/library.cairo [hunk 10]
```diff
@@ -66,6 +66,10 @@ func Account_jumpdests_initialized() -> (initialized: felt) {
 func Account_authorized_message_hashes(hash: Uint256) -> (res: felt) {
 }
 
+@storage_var
+func Account_code_hash() -> (codeHash: Uint256) {
```
Comment (enitrat) [inline src/kakarot/accounts/library.cairo:70]:
```suggestion
func Account_code_hash() -> (code_hash: Uint256) {
```

File: tests/src/kakarot/accounts/test_account_contract.py [hunk 11]
```diff
@@ -258,6 +258,21 @@ def test__should_return_if_jumpdest_valid_when_not_stored(
...
 
+    class TestCodeHash:
+        @SyscallHandler.patch("Ownable_owner", 0xDEAD)
+        def test_should_assert_only_owner(self, cairo_run):
+            with cairo_error(message="Ownable: caller is not the owner"):
+                cairo_run("test__set_code_hash", code_hash=int_to_uint256(0x32))
+
+        @SyscallHandler.patch("Ownable_owner", SyscallHandler.caller_address)
+        def test__should_set_code_hash(self, cairo_run):
+            cairo_run("test__set_code_hash", code_hash=int_to_uint256(0x32))
```
Comment (obatirou) [inline tests/src/kakarot/accounts/test_account_contract.py:269]: Add fuzzing

File: tests/src/kakarot/accounts/test_account_contract.py [hunk 12]
```diff
@@ -258,6 +258,21 @@ def test__should_return_if_jumpdest_valid_when_not_stored(
...
+
+            code_hash = get_storage_var_address("Account_code_hash")
+            SyscallHandler.mock_storage.assert_has_calls(
+                [call(address=code_hash, value=0x32)]
+            )
```
Comment (enitrat) [inline tests/src/kakarot/accounts/test_account_contract.py:274]: can also use

```suggestion
            SyscallHandler.mock_storage.assert_called_with(
                address=code_hash,
                value=0x32,
            )
```

File: src/kakarot/account.cairo (hunk 7 above)
Comment (ClementWalter) [reply src/kakarot/account.cairo:144]: note: I suggested this. In some case I don't manage to use `&` and this is a workaround

File: src/kakarot/instructions/system_operations.cairo [hunk 13]
```diff
@@ -1205,9 +1205,13 @@ namespace CreateHelper {
...
         }
 
-        // Write bytecode and valid jumpdests to Account
+        // Write bytecode, valid jumpdests and code_hash to Account
         let account = State.get_account(evm.message.address.evm);
         let account = Account.set_code(account, evm.return_data_len, evm.return_data);
+        Account.compute_code_hash(account);
+        let (ap_val) = get_ap();
+        let code_hash = cast(ap_val - 1, Uint256*);
+        let account = Account.set_code_hash(account, code_hash);
```
Comment (ClementWalter) [inline src/kakarot/instructions/system_operations.cairo]: we could consider computing the code_hash when we set the code in `Account.set_code` instead of creating another `Account.set_code_hash` indeed, because it creates the risk to have a discrepency between code and code_hash

File: src/kakarot/interpreter.cairo [hunk 14]
```diff
@@ -1047,6 +1047,10 @@ namespace Internals {
         // Write bytecode and cache the final code valid jumpdests to Account
         let account = State.get_account(evm.message.address.evm);
         let account = Account.set_code(account, evm.return_data_len, evm.return_data);
+        Account.compute_code_hash(account);
+        let (ap_val) = get_ap();
+        let code_hash = cast(ap_val - 1, Uint256*);
+        let account = Account.set_code_hash(account, code_hash);
```
Comment (ClementWalter) [inline src/kakarot/interpreter.cairo]: see comment above about `Account.set_code` and code_hash computation

File: src/kakarot/instructions/system_operations.cairo (hunk 13 above)
Comment (obatirou) [reply src/kakarot/instructions/system_operations.cairo]: Done should be better

**Review by enitrat — APPROVED** [review]
lgtm

File: src/kakarot/account.cairo [hunk 15]
```diff
@@ -650,6 +689,26 @@ namespace Account {
         );
         return self;
     }
+
+    func compute_code_hash{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
```
Comment (enitrat) [inline src/kakarot/account.cairo:693]: actually - we should have a test for this

**Review by enitrat — CHANGES_REQUESTED** [review]

File: src/kakarot/account.cairo (hunk 15 above)
Comment (obatirou) [reply src/kakarot/account.cairo:693]: Done with [ea35f94](https://github.com/kkrt-labs/kakarot/pull/1309/commits/ea35f9463c2bc38432bf3b7232f26cf726935a4f)

Comment (obatirou) [reply src/kakarot/account.cairo:693]: Fixed with [ed6c1db](https://github.com/kkrt-labs/kakarot/pull/1309/commits/ed6c1db86fb3c1d37c65fb15e34d46275a698652)
`cairo_run` interpret the return as a int and do not take into account the leading 0
https://github.com/kkrt-labs/kakarot/actions/runs/10112754238/job/27967569334

File: src/kakarot/account.cairo [hunk 16]
```diff
@@ -650,6 +689,26 @@ namespace Account {
...
+        code_len: felt, code: felt*
+    ) -> Uint256 {
+        alloc_locals;
```
Comment (enitrat) [inline src/kakarot/account.cairo:696]: if code_len = 0 you should do an early return with the hash of empty bytecode which is per https://eips.ethereum.org/EIPS/eip-1052 `c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470 `. This might be why you have ef-test failing

File: src/kakarot/instructions/system_operations.cairo [hunk 17]
```diff
@@ -1205,7 +1204,7 @@ namespace CreateHelper {
             return evm;
         }
 
-        // Write bytecode and valid jumpdests to Account
+        // Write bytecode, valid jumpdests and code_hash to Account
```
Comment (obatirou) [inline src/kakarot/instructions/system_operations.cairo:1207]: revert this change

**Comment by obatirou** [comment]
Blocked until a PR is done in EF-test to set the code hash in pre state.
